### PR TITLE
[POC] Add ACK for And Group/Leave Group/Send Group

### PR DESCRIPTION
**Pull request by zackliu**
A poc only apply ack on .net core signalr

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/Utilities/AckHandler.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    internal class AckHandler : IDisposable
+    {
+        private readonly ConcurrentDictionary<string, AckInfo> _acks = new ConcurrentDictionary<string, AckInfo>();
+        private readonly Timer _timer;
+        private readonly TimeSpan _ackThreshold = TimeSpan.FromSeconds(10);
+        private readonly TimeSpan _ackInterval = TimeSpan.FromSeconds(5);
+
+        public AckHandler()
+        {
+            _timer = new Timer(_ => CheckAcks(), state: null, dueTime: _ackInterval, period: _ackInterval);
```
Comment (BrennanConroy) [inline src/Microsoft.Azure.SignalR.Common/Utilities/AckHandler.cs]: Consider not capturing the ExecutionContext, see https://github.com/aspnet/AspNetCore/blob/master/src/SignalR/server/StackExchangeRedis/src/Internal/AckHandler.cs#L22

File: src/Microsoft.Azure.SignalR.Protocols/GroupMessage.cs [hunk 2]
```diff
@@ -56,4 +56,86 @@ public LeaveGroupMessage(string connectionId, string groupName)
...
+        /// <param name="ackId">The ack Id</param>
+        public LeaveGroupWithAckMessage(string connectionId, string groupName, string ackId = null)
+        {
+            ConnectionId = connectionId;
+            GroupName = groupName;
+            AckId = ackId;
+        }
+    }
+
+    public class GroupAckMessage : ServiceMessage, IAckableMessage
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Protocols/GroupMessage.cs]: Ack message?

File: src/Microsoft.Azure.SignalR.Common/Utilities/AckHandler.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                    ack.Tcs.TrySetCanceled();
+                }
+            }
+        }
+
+        private class AckInfo
+        {
+            public TaskCompletionSource<object> Tcs { get; private set; }
+
+            public DateTime Created { get; private set; }
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/Utilities/AckHandler.cs]: Created => Expired?

File: src/Microsoft.Azure.SignalR.Common/Utilities/AckHandler.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            }
+        }
+
+        public void Dispose()
+        {
+            _timer?.Dispose();
+
+            foreach (var pair in _acks)
+            {
+                if (_acks.TryRemove(pair.Key, out var ack))
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/Utilities/AckHandler.cs]: no need?

File: src/Microsoft.Azure.SignalR.Protocols/GroupMessage.cs [hunk 5]
```diff
@@ -56,4 +56,86 @@ public LeaveGroupMessage(string connectionId, string groupName)
...
+    {
+        /// <summary>
+        /// Gets or sets the ack id.
+        /// </summary>
+        public string AckId { get; set; }
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Protocols/GroupMessage.cs]: should we use int32?

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs [hunk 6]
```diff
@@ -118,6 +118,23 @@ public Task WriteAsync(ServiceMessage serviceMessage)
...
+            }
+
+            var routed = GetRoutedEndpoints(serviceMessage, _endpointManager.GetAvailableEndpoints()).ToArray();
+
+            if (routed.Length == 0)
+            {
+                throw new AzureSignalRNotConnectedException();
+            }
+
+            return Task.WhenAny(routed.Select(s => Connections[s]).Select(s => s.WriteAndWaitForAckAsync(serviceMessage)));
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs]: WhenAll?

File: src/Microsoft.Azure.SignalR.Common/Utilities/AckHandler.cs (hunk 4 above)
Comment (BrennanConroy) [reply src/Microsoft.Azure.SignalR.Common/Utilities/AckHandler.cs]: It could be removed by `CheckAcks` so should keep it.

**Comment by zackliu** [comment]
This POC will be splited to multiple PR to impliement 
Add protocols --> #390

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs (hunk 6 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs]: `WhenAny` for `GroupJoin` and `GroupLeave` success, or `WhenAll` fail

**Pull request closed without merging**